corehq/apps/reports/standard/users/reports.py

```diff
@@ -99,6 +128,10 @@ def _build_query(self, user_ids, changed_by_user_ids, actions, user_upload_recor
         if changed_by_user_ids:
             filters = filters & Q(changed_by__in=changed_by_user_ids)
 
+        if property:
+            key = f"details__changes__{property}__isnull"
```

Comment:
this field isn't indexed, is that going to be a performance concern?

Reply:
Also, might be worth making a note in the model class about the expected structure of the JSON field.

Reply:
74af9399b2c - I don't really know if it'll be a performance concern, I'm just kind of assuming that this table will get quite large and that searching a nested JSON field will be slow. If you have any general thoughts on when to add indices, I'm interested.

Reply:
can you confirm if that new index is actually used when we run this query?

Reply:
Also, I think it would be better to use `has_key`: https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/fields/#std:fieldlookup-hstorefield.has_key

```
Q(details__changes__has_key=property)
```

(Would just need to confirm that it works for keys that are nested)

Reply:
Updated to use `has_key` in 88295f557a4 and verified it works.

> can you confirm if that new index is actually used when we run this query?

Good call. It is not used. Docs suggest a GIN index would be the way to best index this field, but it looks like it would need a separate index for each individual field. That's 8 new indices that would get updated every time a user is saved. Given that bulk user uploads are already a performance issue and that this report is new and we don't know which filtering will be most valuable, I'm inclined to not add any indexes at this point. ed70e2ff984 removes the unhelpful B-tree index I had added. Open to feedback.

Reply:
+1 to reverting the index and see how much this hits performance.

confession, Searching by each property was not something I expected to come on this report, hence it was not structured to be searchable.

Reply:
You would only need a single GIN index which would cover the whole field: https://www.postgresql.org/docs/current/datatype-json.html#JSON-INDEXING

It shouldn't be hard to create one manually in a Django migration.

Reply:
Hmm, so to experiment with this I created this index locally in `dbshell`:

`create index users_userh_details_idx on users_userhistory using gin (details);`

But when querying, it doesn't seem to get used. These queries all still do sequential scans:

```
In [12]: UserHistory.objects.filter(details__isnull=False).values("details").explain()
Out[12]: 'Seq Scan on users_userhistory  (cost=0.00..1.06 rows=6 width=32)\n  Filter: (details IS NOT NULL)'

In [13]: UserHistory.objects.filter(details__changes__isnull=False).values("details").explain()
Out[13]: "Seq Scan on users_userhistory  (cost=0.00..1.07 rows=6 width=32)\n  Filter: ((details -> 'changes'::text) IS NOT NULL)"

In [14]: UserHistory.objects.filter(details__changes__contains="a").values("details").explain()
Out[14]: 'Seq Scan on users_userhistory  (cost=0.00..1.09 rows=1 width=32)\n  Filter: ((details -> \'changes\'::text) @> \'"a"\'::jsonb)'
```

I can look more tomorrow, but curious if I"m doing something obviously wrong.

Reply:
The Django queries aren't using the supported index operations:

> The default GIN operator class for jsonb supports queries with top-level key-exists operators ?, ?& and ?| operators and path/value-exists operator @>.

From the Django docs the query functions that are available are:
* contains (`@>`)
  * you did test this but I think you'd need to pass a dict instead of just a key, also I tested this below and it doesn't work for nested objects. https://www.postgresql.org/docs/current/datatype-json.html#JSON-CONTAINMENT
* has_key (`?`)
* has_keys (`?&`)
* has_any_keys (`?|`)

I think the only question is whether or not the index can be used for nested keys. The docs seems to say no and I've confirmed that with a quick shell test:
```sql
-- we can find 'c' because it's a top-level key
select '{"c": {"d": 3}}'::jsonb ? 'c'; -- yields true

-- we can't find 'd' because it's not a top-level key
select '{"c": {"d": 3}}'::jsonb ? 'd'; -- yields false

-- we also can't check for containment since it doesn't support rested objects (and we'd need to know the value)
SELECT '{"foo": {"bar": "baz"}}'::jsonb @> '{"bar": "baz"}'::jsonb;  -- yields false
```

So with the current structure of `details` there is no way to efficiently do the query you are trying to do.

Reply:
Made some notes in the design doc (https://docs.google.com/document/d/15WS2OuB8l8UJePuwnZASIerxIrB2Gfnf1RRt24O3n8M/edit?disco=AAAANIegoeI) but just noting here as well that it may make sense to put the list of changed fields in a separate DB field for the purpose of efficient querying.

Reply:
That makes sense, thank you for the research.